torch/csrc/jit/codegen/cuda/lower_predicate.cpp

```diff
@@ -175,6 +177,492 @@ std::vector<kir::Expr*> generateConditionalFromPredicate(
   return mutated_exprs;
 }
 
+namespace {
+
+//! Analyze whether IterDomain can be statically determined to be safe
+//! without bounds-checking predicates.
+//! TODO: Merge this with PredicateElimination
+class IterationDomainAnalysis : private OptOutDispatch {
+ public:
+  //! Return true if the expression defining tv can be safely run
+  //! without a predicate
+  static bool canOmitPredicate(const TensorDomain* td) {
+    const auto gpu_lower = GpuLower::current();
+    for (size_t i = 0; i < td->nDims(); ++i) {
+      IterDomain* id = gpu_lower->caLoopMap().getConcreteMappedID(td->axis(i));
+      IterationDomainAnalysis id_analysis(id->fusion());
+      auto extent = id->extent();
+      id_analysis.handle(extent);
+      if (!id_analysis.isExact(extent)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+ private:
+  IterationDomainAnalysis(Fusion* fusion) : fusion_(fusion) {}
+
+  using OptOutDispatch::handle;
+
+  //! Check if val has nothing that prevents a loop using val as its
+  //! extent to omit a bounds-checking predicate
+  bool isExact(const Val* val) {
+    return exact_vals_.find(val) != exact_vals_.end();
+  }
+
+  //! Record val does not need a predicate.
+  void setExact(const Val* val) {
+    exact_vals_.insert(val);
+  }
+
+  void handle(Val* val) override {
+    if (val->definition() != nullptr) {
+      handle(val->definition());
+    } else {
+      setExact(val);
+    }
+  }
+
+  void handle(BinaryOp* bop) override {
+    const auto lhs = bop->lhs();
+    const auto rhs = bop->rhs();
+
+    handle(lhs);
+    handle(rhs);
+
+    if (!(isExact(lhs) && isExact(rhs))) {
+      return;
+    }
+
+    if (bop->getBinaryOpType() == BinaryOpType::CeilDiv) {
+      // CeilDiv is the only expression that can make an extent val
+      // larger than the actual. Need to know the exact values.
+      ExpressionEvaluator ee(fusion_);
+      const auto lhs_value = ee.evaluate(lhs);
+      const auto rhs_value = ee.evaluate(rhs);
+      if (lhs_value.has_value() && rhs_value.has_value() &&
+          (lhs_value.value() % rhs_value.value()) == 0) {
+        setExact(bop->out());
+      }
+    } else if (bop->getBinaryOpType() == BinaryOpType::Mul) {
+      setExact(bop->out());
+    } else {
+      // Expr on extent should be either CeilDiv or Mul, which are
+      // derived from split and merge, respectively.
+      TORCH_INTERNAL_ASSERT("Unexpected BinaryOpType: ", bop);
+    }
+  }
+
+ private:
+  Fusion* fusion_ = nullptr;
+  //! Vals that are known to need no predicate if used as IterDomain extent
+  std::unordered_set<const Val*> exact_vals_;
+};
```

Comment:
Moved from predicate_compute.cpp. I'll merge this with `PredicateAnalyzer` as a follow-up refactoring PR.